### PR TITLE
Added category entity to the project

### PR DESCRIPTION
**Pull request by kaiiiseeel**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/entity/Book.java [hunk 1]
```diff
@@ -38,4 +44,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (yevheniisynytsia) [inline src/main/java/mate/academy/bookstore/entity/Book.java]:
```suggestion
    @ManyToMany(fetch = FetchType.LAZY)
```

File: src/main/java/mate/academy/bookstore/entity/Book.java [hunk 2]
```diff
@@ -38,4 +44,11 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted = false;
+    @ManyToMany(fetch = FetchType.EAGER)
+    @JoinTable(
+            name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id")
+    )
+    private Set<Category> categories = new HashSet<>();
```
Comment (yevheniisynytsia) [inline src/main/java/mate/academy/bookstore/entity/Book.java:55]: please exclude from equals hashcode tostring

File: src/main/java/mate/academy/bookstore/entity/Category.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    private Long id;
+    @Column(nullable = false, unique = true)
+    private String name;
+    private String description;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+}
+
+
+
```
Comment (yevheniisynytsia) [inline src/main/java/mate/academy/bookstore/entity/Category.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/mapper/CategoryMapper.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import mate.academy.bookstore.config.MapperConfig;
+import mate.academy.bookstore.dto.CategoryRequestDto;
+import mate.academy.bookstore.dto.CategoryResponseDto;
+import mate.academy.bookstore.entity.Category;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface CategoryMapper {
+
```
Comment (yevheniisynytsia) [inline src/main/java/mate/academy/bookstore/mapper/CategoryMapper.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    public CategoryResponseDto save(CategoryRequestDto categoryRequestDto) {
+        Category category = categoryMapper.toModel(categoryRequestDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public CategoryResponseDto update(Long id, CategoryRequestDto categoryRequestDto) {
+        if (!categoryRepository.existsById(id)) {
+            throw new EntityNotFoundException("Can't update category with id: " + id);
+        }
```
Comment (yevheniisynytsia) [inline src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java]: if is duplicated, can be extracted into a separate method

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]
Well done! Left some suggestions

File: src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        categoryRepository.deleteById(id);
+
+    }
+
+    private boolean isCategoryExists(Long id) {
+        return categoryRepository.existsById(id);
+    }
+}
+
+
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/service/impl/CategoryServiceImpl.java]:
```suggestion

```

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -19,28 +20,28 @@ public class BookServiceImpl implements BookService {
     private final BookMapper bookMapper;
 
     @Override
-    public BookDto save(BookRequestDto requestDto) {
+    public BookResponseDto save(BookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
         return bookMapper.toDto(bookRepository.save(book));
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java:34]: You need to save and validate categories as well here

File: src/main/java/mate/academy/bookstore/repository/BookRepository.java [hunk 8]
```diff
@@ -1,8 +1,14 @@
...
 import mate.academy.bookstore.entity.Book;
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
 public interface BookRepository extends JpaRepository<Book, Long> {
 
+    @Query("FROM Book b JOIN FETCH b.categories c WHERE c.id = :categoryId")
+    List<Book> findAllByCategoryId(Long categoryId, Pageable pageable);
+
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/repository/BookRepository.java]:
```suggestion

```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job! Left some comments

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 9]
```diff
@@ -17,30 +23,42 @@
...
+        Set<Long> categoryIds = requestDto.getCategoryIds();
+        List<Category> categories = categoryRepository.findAllById(categoryIds);
+
+        if (categoryIds.size() != categories.size()) {
+            Set<Long> availableCategories = categories.stream()
+                    .map(Category::getId)
+                    .collect(Collectors.toSet());
+            categoryIds.removeAll(availableCategories);
+            throw new EntityNotFoundException("Category id not found: " + categoryIds);
+        }
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java]: Move thos to separate method

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -17,30 +23,42 @@
...
-    public BookDto save(BookRequestDto requestDto) {
+    public BookResponseDto save(BookRequestDto requestDto) {
+        Set<Long> categoryIds = requestDto.getCategoryIds();
+        List<Category> categories = categoryRepository.findAllById(categoryIds);
+
+        if (categoryIds.size() != categories.size()) {
+            Set<Long> availableCategories = categories.stream()
+                    .map(Category::getId)
+                    .collect(Collectors.toSet());
+            categoryIds.removeAll(availableCategories);
```
Comment (andrii-hoienko) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java]: What is the reason to make some removal logic, all you need is compare size to realize that all categories are present

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left some comments

**Pull request merged**